src/unix/pipe.c

```diff
@@ -287,6 +287,58 @@ int uv_pipe_getpeername(const uv_pipe_t* handle, char* buffer, size_t* size) {
 }
 
 
+int uv_pipe_get_credentials(const uv_pipe_t* handle, uv_credentials_t* creds) {
+  int r;
+  uv_os_fd_t os_fd;
+
+  r = uv_fileno((uv_handle_t*) handle, &os_fd);
+  if (r != 0)
+    return r;
+
+#if defined(__APPLE__)
+  {
+    // https://stackoverflow.com/a/38207320
+    unsigned int uid;
+    unsigned int gid;
+    socklen_t pid_len = sizeof(creds->pid);
+
+    r = getsockopt(os_fd,
+                   SOCK_STREAM,
+                   LOCAL_PEERPID,
+                   &creds->pid,
+                   &pid_len);
```

Comment:
There is a better/atomic way to get the credentials: `LOCAL_PEERCRED` with a `struct xucred` as the argument. Include `<sys/ucred.h>` to get its definition.

Reply:
Unfortunately `struct xucred` doesn't include the PID. Somewhat relevant: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=215202

Should I still replace the call to `getpeereid()`? I'm not sure which one is preferred.

Reply:
@bnoordhuis poke ^-^

Reply:
Ah, you're right. Okay, `LOCAL_PEERCRED` isn't going to help then.

I wonder if it matters that there's a time window between the two system calls. I can't think of a credible exploit but maybe I'm simply not imaginative enough.

Reply:
I think part of the point of them being non-forgeable is that there's no TOCTOU-type exploits from reading the information in separate syscalls. Relatedly though, it's worthwhile being aware than any use of `pid` directly is vulnerable to a TOCTOU-type exploit (as mentioned in that bug report https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=215202#c2). But use of uid is safe.

Thus, on Windows, where the user might be tempted to use the `pid` to lookup the credentials that does probably present a security vulnerable of this sort (e.g. malicious user launches a process, get's libuv to read the pid, then kills the process and launches SYSTEM applications until one runs with the same pid and thereby tricking libuv into giving the wrong privileges to a user process)

The MacOS docs say that `getpeereid` is simply a wrapper for `LOCAL_PEERCRED`, so `getpeereid` seems good to me.